README.md
<p align="center">
    <img src="https://github.com/yelp/MOE/raw/master/moe/static/img/MOE_full_logo.png" alt="MOE Logo">
</p>

[![Build Status](https://travis-ci.org/Yelp/MOE.svg?branch=master)](https://travis-ci.org/Yelp/MOE)

Metric Optimization Engine. A global, black box optimization engine for real world metric optimization.

  * [Full documentation][1]
  * [REST documentation][2]

Or, build the documentation locally with `make docs`.

## What is MOE?

MOE (Metric Optimization Engine) is an *efficient* way to optimize a system's parameters, when evaluating parameters is *time-consuming* or *expensive*.

Here are some examples of when you could use MOE:

* **Optimizing a system's click-through rate (CTR).**  MOE is useful when evaluating CTR requires running an A/B test on real user traffic, and getting statistically significant results requires running this test for a substantial amount of time (hours, days, or even weeks).

* **Optimizing tunable parameters of a machine-learning prediction method.**  MOE is useful if calculating the prediction error for one choice of the parameters takes a long time, which might happen because the prediction method is complex and takes a long time to train, or because the data used to evaluate the error is huge.

* **Optimizing the design of an engineering system** (an airplane, the traffic network in a city, a combustion engine, a hospital).  MOE is useful if evaluating a design requires running a complex physics-based numerical simulation on a supercomputer. 

* **Optimizing the parameters of a real-world experiment** (a chemistry, biology, or physics experiment, a drug trial).  MOE is useful when every experiment needs to be physically created in a lab, or very few experiments can be run in parallel.

MOE is ideal for problems in which the optimization problem's objective function is a black box, not necessarily convex or concave, derivatives are unavailable, and we seek a global optimum, rather than just a local one. This ability to handle black-box objective functions allows us to use MOE to optimize nearly any system, without requiring any internal knowledge or access. To use MOE, we simply need to specify some [objective function][14], some set of [parameters][15], and any historical data we may have from previous evaluations of the objective function. MOE then finds the set of parameters that maximize (or minimize) the objective function, while evaluating the objective function as little as possible. 

Inside, MOE uses *Bayesian global optimization*, which performs optimization using Bayesian statistics and *optimal learning*. 

Optimal learning is the study of efficient methods for collecting information, particularly when doing so is time-consuming or expensive, and was developed and popularized from its roots in decision theory by [Prof. Peter Frazier][16] ([Cornell, Operations Research and Information Engineering][17]) and [Prof. Warren Powell][18] ([Princeton, Operations Research and Financial Engineering][19]). For more information about the mathematics of optimal learning, and more real-world applications like heart surgery, drug discovery, and materials science, see these [intro slides][20] to optimal learning.

[Why do we need MOE?][21]

Video and slidedeck introduction to MOE:

* [15 min MOE intro video][10]
* [MOE intro slides][11]
* [Full documentation][1]

MOE does this internally by:

1. Building a Gaussian Process (GP) with the historical data
2. Optimizing the hyperparameters of the Gaussian Process (model selection)
3. Finding the points of highest Expected Improvement (EI)
4. Returning the points to sample, then repeat

Externally you can use MOE through:

1. [The REST interface][2]
2. [The Python interface][9]
3. [The C++ interface][12]

You can be up and optimizing in a matter of minutes. [Examples of using MOE][13]

# Install

## Install in docker:

This is the recommended way to run the MOE REST server. All dependencies and building is done automatically and in an isolated container.

[Docker (http://docs.docker.io/)][6] is a container based virtualization framework. Unlike traditional virtualization Docker is fast, lightweight and easy to use. Docker allows you to create containers holding all the dependencies for an application. Each container is kept isolated from any other, and nothing gets shared.

```bash
$ docker pull yelpmoe/latest # You can also pull specific versions like yelpmoe/v0.1.0
$ docker run -p 6543:6543 yelpmoe/latest
```
If you are on OSX, or want a build based on the current master branch you may need to build this manually.

```bash
$ git clone https://github.com/Yelp/MOE.git
$ cd MOE
$ docker build -t moe_container .
$ docker run -p 6543:6543 moe_container
```

The webserver and REST interface is now running on port 6543 from within the container. http://localhost:6543

## Install from source:

See [Install Documentation][7]

# Running MOE

## REST/web server and interactive demo

from the directory MOE is installed:

```bash
$ pserve --reload development.ini # MOE server is now running at http://localhost:6543
```

[The REST interface documentation][2]

Or, from the command line,

```bash
$ curl -X POST -H "Content-Type: application/json" -d '{"domain_info": {"dim": 1}, "points_to_evaluate": [[0.1], [0.5], [0.9]], "gp_historical_info": {"points_sampled": [{"value_var": 0.01, "value": 0.1, "point": [0.0]}, {"value_var": 0.01, "value": 0.2, "point": [1.0]}]}}' http://127.0.0.1:6543/gp/ei
```
[`gp_ei` endpoint documentation.][4]

## From ipython

```bash
$ ipython
> from moe.easy_interface.experiment import Experiment
> from moe.easy_interface.simple_endpoint import gp_next_points
> exp = Experiment([[0, 2], [0, 4]])
> exp.historical_data.append_sample_points([[[0, 0], 1.0, 0.01]])
> next_point_to_sample = gp_next_points(exp)
> print next_point_to_sample
```
[`easy_interface` documentation.][5]

## Within Python

See ``examples/next_point_via_simple_endpoint.py`` for this code or http://yelp.github.io/MOE/examples.html for more examples.

```python
import math
import random

from moe.easy_interface.experiment import Experiment
from moe.easy_interface.simple_endpoint import gp_next_points
from moe.optimal_learning.python.data_containers import SamplePoint


# Note: this function can be anything, the output of a batch, results of an A/B experiment, the value of a physical experiment etc.
def function_to_minimize(x):
    """Calculate an aribitrary 2-d function with some noise with minimum near [1, 2.6]."""
    return math.sin(x[0]) * math.cos(x[1]) + math.cos(x[0] + x[1]) + random.uniform(-0.02, 0.02)

if __name__ == '__main__':
    exp = Experiment([[0, 2], [0, 4]])  # 2D experiment, we build a tensor product domain
    # Bootstrap with some known or already sampled point(s)
    exp.historical_data.append_sample_points([
        SamplePoint([0, 0], function_to_minimize([0, 0]), 0.05),  # Iterables of the form [point, f_val, f_var] are also allowed
        ])

    # Sample 20 points
    for i in range(20):
        # Use MOE to determine what is the point with highest Expected Improvement to use next
        next_point_to_sample = gp_next_points(exp)[0]  # By default we only ask for one point
        # Sample the point from our objective function, we can replace this with any function
        value_of_next_point = function_to_minimize(next_point_to_sample)

        print "Sampled f({0:s}) = {1:.18E}".format(str(next_point_to_sample), value_of_next_point)

        # Add the information about the point to the experiment historical data to inform the GP
        exp.historical_data.append_sample_points([SamplePoint(next_point_to_sample, value_of_next_point, 0.01)])  # We can add some noise
```

More examples can be found in the `<MOE_DIR>/examples` directory.

## Within C++

Expected Improvement Demo - http://yelp.github.io/MOE/gpp_expected_improvement_demo.html
Gaussian Process Hyperparameter Optimization Demo - http://yelp.github.io/MOE/gpp_hyperparameter_optimization_demo.html
Combined Demo - http://yelp.github.io/MOE/gpp_hyper_and_EI_demo.html

# Contributing

See [Contributing Documentation][8]

# License

MOE is licensed under the Apache License, Version 2.0: http://www.apache.org/licenses/LICENSE-2.0

[0]: https://www.youtube.com/watch?v=CC6qvzWp9_A
[1]: http://yelp.github.io/MOE/
[2]: http://yelp.github.io/MOE/moe.views.rest.html
[3]: http://github.com/Yelp/MOE/pulls
[4]: http://yelp.github.io/MOE/moe.views.rest.html#module-moe.views.rest.gp_ei
[5]: http://yelp.github.io/MOE/moe.easy_interface.html
[6]: http://docs.docker.io/
[7]: http://yelp.github.io/MOE/install.html
[8]: http://yelp.github.io/MOE/contributing.html
[9]: http://yelp.github.io/MOE/moe.optimal_learning.python.python_version.html
[10]: http://www.youtube.com/watch?v=CC6qvzWp9_A
[11]: http://www.slideshare.net/YelpEngineering/optimal-learning-for-fun-and-profit-with-moe
[12]: http://yelp.github.io/MOE/cpp_tree.html
[13]: http://yelp.github.io/MOE/examples.html
[14]: http://yelp.github.io/MOE/objective_functions.html
[15]: http://yelp.github.io/MOE/objective_functions.html#parameters
[16]: http://people.orie.cornell.edu/pfrazier/
[17]: http://www.orie.cornell.edu/
[18]: http://optimallearning.princeton.edu/
[19]: http://orfe.princeton.edu/
[20]: http://people.orie.cornell.edu/pfrazier/Presentations/2014.01.Lancaster.BGO.pdf
[21]: http://yelp.github.io/MOE/why_moe.html


CHANGELOG.md
* Features

* Changes

* Bugs

## v0.2.2 (2014-12-5)

SHA: ``83a71512e6a4aa553866e4301ef11f172d485e23``

* Features

  * Added startup message to REST server including tips for OSX users (#400)
  * Added GPU support to ``cpp_wrappers.expected_improvement.multistart_expected_improvement_optimization``; requires ``max_num_threads == 1`` until future multi-GPU support (#368)
  * Added the COBYLA optimizer to the expected improvement optimization class. (#370)
  * OptimizerParameter struct members now directly readable/writeable from Python; added EqualityComparisonMixin (#138)
  * C++ GradientDescentParameters object now stores ``num_steps_averaged`` (but does not use it yet) (#391)
  * conda build system for MOE (#417)

* Changes

  * Switched from `testify` to `py.test` - http://pytest.org/ (#36)
  * [cleanup] Moved bandits into their own sub directories (#375)
  * Supply PYTHON_LIBRARY and PYTHON_INCLUDE_DIR vars to cmake automatically in ``setup.py`` (#412)
  * Added warning when colander version is out of date (#413)

* Bugs

  * Fixed UCB1 and UCB1-tuned algorithm confidence bound calculation (#432)

## v0.2.1 (2014-09-22)

SHA: ``ab6f959c11a0cacbed6dad618fe6ffed71092116``

* Features

  * Implemented BLA (Bayesian Learning Automaton). (#373)
  * Connected GPU functions to multistart gradient descent optimizer. (#270)

* Changes

* Bugs

  * variance in a sample arm was dropped in _make_bandit_historical_info_from_params. (#385)
  * SampleArm's ``__add__`` and ``__str__`` were broken. (#387)
  * Specifying ``max_num_threads`` on GPU compute paths caused a segfault (#394)

## v0.2.0 (2014-08-15)

SHA: ``8201917e3f9b47b8edd8039ea3278ef8631b0f2a``

* Features

  * Added multi-armed bandit endpoint. (#255)
    * Implemented epsilon-greedy. (#255)
    * Implemented epsilon-first. (#335)
    * Implemented UCB1. (#354)
    * Implemented UCB1-tuned. (#366)
  * Added support for the L-BFGS-B optimizer. (#296)
  * Added GPU implementation for q,p-EI and its gradient computation. (#219)
    * Speed up GPU functions by redesign of memory allocation. (#297)

* Changes

  * Split up old ``schemas.py`` file into ``schemas/`` directory with several subfiles (#291)
  * Improved Dockerfile, reducing Docker-based install times substantially, https://hub.docker.com/u/yelpmoe/ (#332)
    * Created ``min_reqs`` docker container which is a snapshot of all MOE third-party requirements
    * Created ``latest``, which tracks the latest MOE build
    * Started releasing docker containers for each tagged MOE release (currently just ``v0.1.0``)
  * ``GradientDescentOptimization`` (C++) no longer has a separate ``next_points`` output (#186)
  * LogLikelihood evaluate at point list and latin hypercube search now return status dicts like every other optimizer (#189)
    * status dicts also a little more informative/standardized now
  * Update C++ autodoc tools to handle the new ``gpu`` directory (#353)
  * Added ``__version__`` to ``moe/__init__.py`` (#353)

* Bugs

  * Throw exceptions (C++) if ``num_multistarts`` or ``num_random_samples`` is 0 (#345)
  * ``combined_example`` endpoint was not passing ``kwargs`` through so users could not change the default server (#356)
    * fix sometimes dropped general ``kwargs`` (#358)
  * ``mean_var_of_gp_from_historic_data`` was also not passing ``kwargs`` (#359)

## v0.1.0 (2014-07-29)

SHA: ``5fef1d242cc8b6e0d6443522f8ba73ba743607de``

* Features

  * initial open source release


moe/views/gp_ei.py
# -*- coding: utf-8 -*-
"""Classes for gp_ei endpoints.

Includes:
    1. request and response schemas
    2. pretty and backend views
"""
import colander
import numpy
from pyramid.view import view_config

from moe.optimal_learning.python.constant import default_expected_improvement_parameters
from moe.views.gp_pretty_view import GpPrettyView, PRETTY_RENDERER
from moe.views.schemas import ListOfPointsInDomain, GpInfo, ListOfExpectedImprovements
from moe.views.utils import _make_gp_from_gp_info
from moe.views.constant import GP_EI_ROUTE_NAME, GP_EI_PRETTY_ROUTE_NAME


class GpEiRequest(colander.MappingSchema):

    """A gp_ei request colander schema.

    **Required fields**

        :points_to_evaluate: list of points in domain to calculate Expected Improvement (EI) at (moe.views.schemas.ListOfPointsInDomain)
        :gp_info: a moe.views.schemas.GpInfo object of historical data

    **Optional fields**

        :points_being_sampled: list of points in domain being sampled (default: []) (moe.views.schemas.ListOfPointsInDomain)
        :mc_iterations: number of Monte Carlo (MC) iterations to perform in numerical integration to calculate EI (default: 1000)

    **Example Request**

    .. sourcecode:: http

        Content-Type: text/javascrip

        {
            'points_to_evaluate': [[0.1], [0.5], [0.9]],
            'points_being_sampled': [],
            'mc_iterations': 1000,
            'gp_info': {
                'points_sampled': [
                        {'value_var': 0.01, 'value': 0.1, 'point': [0.0]},
                        {'value_var': 0.01, 'value': 0.2, 'point': [1.0]}
                    ],
                'domain': [
                    [0, 1],
                    ]
                },
            },
        }

    """

    points_to_evaluate = ListOfPointsInDomain()
    points_being_sampled = ListOfPointsInDomain(
            missing=[],
            )
    mc_iterations = colander.SchemaNode(
            colander.Int(),
            validator=colander.Range(min=1),
            missing=default_expected_improvement_parameters.mc_iterations,
            )
    gp_info = GpInfo()


class GpEiResponse(colander.MappingSchema):

    """A gp_ei response colander schema.

    **Output fields**

        :endpoint: the endpoint that was called
        :expected_improvement: list of calculated expected improvements (moe.views.schemas.ListOfExpectedImprovements)

    **Example Response**

    .. sourcecode:: http

        {
            "endpoint":"gp_ei",
            "expected_improvement":["0.197246898375","0.443163755117","0.155819546878"]
        }

    """

    endpoint = colander.SchemaNode(colander.String())
    expected_improvement = ListOfExpectedImprovements()


class GpEiView(GpPrettyView):

    """Views for gp_ei endpoints."""

    _route_name = GP_EI_ROUTE_NAME
    _pretty_route_name = GP_EI_PRETTY_ROUTE_NAME

    request_schema = GpEiRequest()
    response_schema = GpEiResponse()

    _pretty_default_request = {
            "points_to_evaluate": [
                [0.1], [0.5], [0.9],
                ],
            "gp_info": GpPrettyView._pretty_default_gp_info,
            }

    @view_config(route_name=_pretty_route_name, renderer=PRETTY_RENDERER)
    def pretty_view(self):
        """A pretty, browser interactive view for the interface. Includes form request and response.

        .. http:get:: /gp/ei/pretty

        """
        return self.pretty_response()

    @view_config(route_name=_route_name, renderer='json', request_method='POST')
    def gp_ei_view(self):
        """Endpoint for gp_ei POST requests.

        .. http:post:: /gp/ei

           Calculates the Expected Improvement (EI) of a set of points, given historical data.

           :input: moe.views.gp_ei.GpEiRequest()
           :output: moe.views.gp_ei.GpEiResponse()

           :status 200: returns a response
           :status 500: server error

        """
        params = self.get_params_from_request()

        points_to_evaluate = numpy.array(params.get('points_to_evaluate'))
        points_being_sampled = numpy.array(params.get('points_being_sampled'))
        gp_info = params.get('gp_info')

        GP = _make_gp_from_gp_info(gp_info)

        expected_improvement = GP.evaluate_expected_improvement_at_point_list(
                points_to_evaluate,
                points_being_sampled=points_being_sampled,
                )

        return self.form_response({
                'endpoint': self._route_name,
                'expected_improvement': expected_improvement.tolist(),
                })
